railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_6_0.rb.tt

```diff
@@ -8,3 +8,10 @@
 
 # Don't force requests from old versions of IE to be UTF-8 encoded
 # Rails.application.config.action_view.default_enforce_utf8 = false
+#
+# Purpose and expiry information are embedded inside signed and encrypted
+# cookies for increased security.
+#
+# This option is not backwards compatible with earlier Rails versions.
+# It's best enabled when your entire app is migrated and stable on 6.0.
+# Rails.application.config.action_dispatch.use_cookies_with_metadata = true
```

Comment:
It's still not clear to me how this meshes with `use_authenticated_cookie_encryption` that's on 5-2-stable: https://github.com/rails/rails/blob/c46a33bf8d661e538f073a348f0d28f4a2129281/actionpack/lib/action_dispatch/middleware/cookies.rb#L493-L500

What should users who upgrade do? And what happens to their cookies, with and without the 5-2-stable flag enabled?

Perhaps you can try some testing on a 5.2 generated app that you upgrade to Rails 6?

Some tips for dev apps:

```
git checkout 5-2-stable # In your Rails checkout
cd ~ # Or wherever you keep your code
path/to/rails/exe/rails new cookie --dev # Generates the app with your Rails checkout (check the Gemfile), then you should be able to do `git checkout your-branch` in Rails.
```

Reply:
@kaspth 
When users upgrade their apps, `use_cookies_with_metadata` is false until they uncomment the config in the new_framework_defaults or load them in `application.config`, given this information, there are two cases here:

i) **5-2-Stable flag is enabled:**
These cookies are written in the new format with expiry and purpose (is nil) embedded. And since purpose checking is off, these cookies are readable on Rails 6.0.

ii) **5-2-Stable flag is disabled:**
These are not AEAD cookies nor do they have expiry and purpose embedded. Since they are converted to AEAD cookies on cookie-read, they are readable too.

For more: https://public.3.basecamp.com/p/pkinaKqD9EM7PtKhtRogD6oQ